### PR TITLE
Remove unnecessary use of ns.follow

### PR DESCRIPTION
**Pull request by tautschnig**
Further cleanup of some remaining unnecessary ns.follow calls after the earlier
removal of ns.follow in these directories.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/ansi-c/c_typecheck_expr.cpp [hunk 1]
```diff
@@ -2875,7 +2875,7 @@ void c_typecheck_baset::typecheck_expr_unary_arithmetic(exprt &expr)
 
   if(o_type.id()==ID_vector)
   {
-    if(is_number(follow(o_type.subtype())))
+    if(is_numbero_type.subtype())
```
Comment (smowton) [inline src/ansi-c/c_typecheck_expr.cpp]: typo

File: src/goto-symex/symex_function_call.cpp [hunk 2]
```diff
@@ -86,24 +86,21 @@ void goto_symext::parameter_assignments(
...
-            f_parameter_type.id()==ID_pointer ||
-            f_parameter_type.id()==ID_union) &&
-           (f_rhs_type.id()==ID_signedbv ||
-            f_rhs_type.id()==ID_unsignedbv ||
-            f_rhs_type.id()==ID_c_bit_field ||
-            f_rhs_type.id()==ID_c_enum_tag ||
-            f_rhs_type.id()==ID_bool ||
-            f_rhs_type.id()==ID_pointer ||
-            f_rhs_type.id()==ID_union))
+        if(
```
Comment (smowton) [inline src/goto-symex/symex_function_call.cpp:94]: Would you mind using `// clang-format off` and keeping the one condition per line? I think this is one of the cases where it really makes a bad call.

**Review by smowton — APPROVED** [review]

File: src/goto-symex/symex_function_call.cpp (hunk 2 above)
Comment (tautschnig) [reply src/goto-symex/symex_function_call.cpp:94]: Indeed - fixed!

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 483d32e001df7a3641e8bac4611c409e443f49ea).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97314753
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7ce01e31972639f28b8c34c383c1bef25c8d9daa).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97317638

File: src/goto-symex/symex_function_call.cpp [hunk 3]
```diff
@@ -86,24 +86,25 @@ void goto_symext::parameter_assignments(
...
+           parameter_type.id() == ID_union_tag) &&
+          (rhs.type().id() == ID_signedbv ||
+           rhs.type().id() == ID_unsignedbv ||
+           rhs.type().id() == ID_c_bit_field ||
+           rhs.type().id() == ID_c_enum_tag ||
+           rhs.type().id() == ID_bool ||
+           rhs.type().id() == ID_pointer ||
+           rhs.type().id() == ID_union ||
+           rhs.type().id() == ID_union_tag))
         {
```
Comment (kroening) [inline src/goto-symex/symex_function_call.cpp:110]: I would retain that alias for rhs.type() -- the .type() method has non-trivial cost.

File: src/pointer-analysis/value_set_dereference.cpp [hunk 4]
```diff
@@ -404,7 +404,7 @@ value_set_dereferencet::valuet value_set_dereferencet::build_reference_to(
 
       result.value=object;
 
-      if(object_type!=ns.follow(dereference_type))
+      if(ns.follow(object_type) != ns.follow(dereference_type))
         result.value.make_typecast(dereference_type);
     }
```
Comment (kroening) [inline src/pointer-analysis/value_set_dereference.cpp:406]: I would remove the .follow() on both sides in the if.
We are going for 'proper type equivalence'.

**Review by kroening — APPROVED** [review]

File: src/cpp/cpp_typecheck_expr.cpp [hunk 5]
```diff
@@ -121,7 +121,6 @@ void cpp_typecheckt::typecheck_expr_main(exprt &expr)
     // these appear to have type "struct _GUID"
     // and they are lvalues!
     expr.type() = struct_tag_typet("tag-_GUID");
-    follow(expr.type());
```
Comment (martin-cs) [inline src/cpp/cpp_typecheck_expr.cpp:124]: Was this always redundant?

File: src/cpp/cpp_typecheck_expr.cpp [hunk 6]
```diff
@@ -415,15 +414,18 @@ bool cpp_typecheckt::overloadable(const exprt &expr)
...
     if(is_reference(t))
       t=t.subtype();
 
-    if(t.id()==ID_struct ||
-       t.id()==ID_union ||
-       t.id()==ID_c_enum || t.id() == ID_c_enum_tag)
+    if(
+      t.id() == ID_struct || t.id() == ID_union || t.id() == ID_c_enum ||
+      t.id() == ID_c_enum_tag || t.id() == ID_struct_tag ||
+      t.id() == ID_union_tag)
```
Comment (martin-cs) [inline src/cpp/cpp_typecheck_expr.cpp:425]: This seems like a bug fix as well...

File: src/pointer-analysis/value_set_fi.cpp [hunk 7]
```diff
@@ -909,13 +909,14 @@ void value_set_fit::get_reference_set_sharing_rec(
...
 
       if(object.id()==ID_unknown)
         insert(dest, exprt(ID_unknown, expr.type()));
-      else if(object.id()==ID_dynamic_object &&
-              obj_type.id()!=ID_struct &&
-              obj_type.id()!=ID_union)
+      else if(
+        object.id() == ID_dynamic_object && obj_type.id() != ID_struct &&
+        obj_type.id() != ID_union && obj_type.id() != ID_struct_tag &&
+        obj_type.id() != ID_union_tag)
```
Comment (martin-cs) [inline src/pointer-analysis/value_set_fi.cpp:919]: Again, a similar bug fix?

**Review by martin-cs — APPROVED** [review]

File: src/cpp/cpp_typecheck_expr.cpp (hunk 5 above)
Comment (tautschnig) [reply src/cpp/cpp_typecheck_expr.cpp:124]: It was introduced in this way in 0b2c6d87 and even back then `follow` didn't have side effects, so I'd say: yes.

File: src/goto-symex/symex_function_call.cpp (hunk 3 above)
Comment (tautschnig) [reply src/goto-symex/symex_function_call.cpp:110]: Done, though I've renamed it from `f_rhs_type` to `rhs_type`.

File: src/pointer-analysis/value_set_dereference.cpp (hunk 4 above)
Comment (tautschnig) [reply src/pointer-analysis/value_set_dereference.cpp:406]: Done, tests appear to pass.

File: src/cpp/cpp_typecheck_expr.cpp (hunk 6 above)
Comment (tautschnig) [reply src/cpp/cpp_typecheck_expr.cpp:425]: Not one I am aware of, other than the `ID_c_enum_tag` having been a dead case before?

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 18ff68386bf02c6426a68c054bba5782017fb02c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97526505

**Pull request merged**